packages/kolme/src/gossip/sync_manager.rs

```diff
@@ -0,0 +1,623 @@
+        match gossip.sync_mode {
+            // State mode allows us to just grab the blocks we need.
+            SyncMode::StateTransfer => (),
+            // Check the archive value and sync from there.
+            SyncMode::Archive => {
+                // Add earlier blocks to archive if we know there are later blocks available.
+                if let Some((next_needed, _)) = self.needed_blocks.first_key_value() {
+                    let next_to_archive = gossip.kolme.get_next_to_archive().await;
+                    if next_to_archive < *next_needed {
+                        self.add_needed_block(gossip, next_to_archive, None).await?;
+                    }
+                }
+            }
+            SyncMode::BlockTransfer => {
+                // If there are any blocks needed that are _later_ than our currently
+                // expected next block, get the next block.
+                if let Some((next_needed, _)) = self.needed_blocks.first_key_value() {
+                    let next_needed = *next_needed;
+                    let next_chain = gossip.kolme.read().get_next_height();
+                    if next_needed != next_chain {
+                        self.add_needed_block(gossip, next_chain, None).await?;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_data_requests_for(
+        gossip: &Gossip<App>,
+        height: BlockHeight,
+        waiting: &mut WaitingBlock<App::Message>,
+    ) -> Result<Option<SmallVec<[DataRequest; REQUEST_COUNT]>>> {
+        let label = DataLabel::Block(height);
+        // Check if it got added to our store in the meanwhile
+        if gossip.kolme.has_block(height).await? {
+            return Ok(None);
+        }
+
+        match waiting {
+            WaitingBlock::Needed(status) => {
+                let request_new_peers = match status.should_request(gossip, label) {
+                    ShouldRequest::DontRequest => {
+                        return Ok(Some(SmallVec::new()));
+                    }
+                    ShouldRequest::RequestNoPeers => false,
+                    ShouldRequest::RequestWithPeers => true,
+                };
+
+                Ok(Some(
+                    std::iter::once(DataRequest {
+                        data: label,
+                        current_peers: status.peers.clone(),
+                        request_new_peers,
+                    })
+                    .collect(),
+                ))
+            }
+            WaitingBlock::Received { block, peer } => {
+                // Determine if we're going to use block or state sync.
+                let do_block = match gossip.sync_mode {
+                    SyncMode::BlockTransfer => true,
+                    SyncMode::StateTransfer | SyncMode::Archive => {
+                        let kolme = gossip.kolme.read();
+                        kolme.get_next_height().next() == block.height()
```

Comment:
Should this have `next()` method call in it ?

Reply:
Oh, you're right, it shouldn't! We need to confirm that this is the next block. I screwed up my own logic. OK, removing.